In [4]:

from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.ensemble import StackingClassifier
from sklearn.svm import LinearSVC
from sklearn.neural_network import MLPClassifier
import pandas as pd
import numpy as np

In [5]:
df=pd.read_csv("C:/Users/ngala/Downloads/insurance.csv")

In [6]:
df.head()


age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

In [7]:
df.shape

(1338, 7)

In [12]:
#Dropping unncessary data
df.drop("bmi",axis="columns",inplace=True)

In [13]:
df.head()

sex  children smoker      charges
0  female         0    yes  16884.92400
1    male         1     no   1725.55230
2    male         3     no   4449.46200
3    male         0     no  21984.47061
4    male         0     no   3866.85520

In [14]:
#Replacing the values in sexr to machine readable
df["sex"].replace({"female":1,"male":0},inplace=True)

In [15]:
df.head()

sex  children smoker      charges
0    1         0    yes  16884.92400
1    0         1     no   1725.55230
2    0         3     no   4449.46200
3    0         0     no  21984.47061
4    0         0     no   3866.85520

In [16]:
#Replacing the values in smokerr to machine readable
df["smoker"].replace({"yes":1,"no":0},inplace=True)

In [17]:
df.head()

sex  children  smoker      charges
0    1         0       1  16884.92400
1    0         1       0   1725.55230
2    0         3       0   4449.46200
3    0         0       0  21984.47061
4    0         0       0   3866.85520

In [18]:
df["charges"]=df.charges
df.head()

sex  children  smoker      charges
0    1         0       1  16884.92400
1    0         1       0   1725.55230
2    0         3       0   4449.46200
3    0         0       0  21984.47061
4    0         0       0   3866.85520

In [19]:
estimators = [
    ('svr', make_pipeline(StandardScaler(),
                           LinearSVC(random_state=42))),
    ('mlp',make_pipeline(StandardScaler(), MLPClassifier(alpha=1, max_iter=100)))
    
 ]
clf = StackingClassifier(
     estimators=estimators, final_estimator=LogisticRegression(random_state=42)
 )

In [21]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
     X, y, stratify=y, random_state=42
 )
clf.fit(X_train, y_train).score(X_test, y_test)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\_multi

0.986013986013986

Above is
Test train split and stratify basd split
calling stacking classifier on train data and score on test data
which gives an accuracy of 98.6 %

In [22]:
svr_clf =make_pipeline(StandardScaler(),
                           LinearSVC(random_state=42))
svr_clf.fit(X_train, y_train).score(X_test, y_test)

0.972027972027972

Above is

LinearSVC has an accuracy of of 97.2


In [23]:
mlp_clf =make_pipeline(StandardScaler(), MLPClassifier(alpha=1, max_iter=100))
mlp_clf.fit(X_train, y_train).score(X_test, y_test)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(


0.9790209790209791

MLPClassifie has an accuracy of 97.9

The stacking classifier has improved the accuracy of the individual classifiers